plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java

```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
             }
 
             Optional<PartitionData> partitionData = getPartitionData(pagePartitioner.getColumns(), page, position);
-            writer = createWriter(partitionData);
+            // prepend query id to a file name so we can determine which files were written by which query. This is needed for opportunistic cleanup of extra files
+            // which may be present for successfully completing query in presence of failure recovery mechanisms.
+            String fileName = fileFormat.toIceberg().addExtension(session.getQueryId() + "-" + randomUUID());
+            String outputPath = partitionData
+                    .map(partition -> locationProvider.newDataLocation(partitionSpec, partition, fileName))
+                    .orElseGet(() -> locationProvider.newDataLocation(fileName));
+
+            ImmutableList.Builder<Integer> sortColumnIndexes = ImmutableList.builder();
+            ImmutableList.Builder<SortOrder> sortOrders = ImmutableList.builder();
+            for (TrinoSortField sortField : sortOrder) {
+                Types.NestedField column = outputSchema.findField(sortField.getSourceColumnId());
+                sortColumnIndexes.add(outputSchema.columns().indexOf(column));
+                sortOrders.add(sortField.getSortOrder());
+            }
+
+            if (!sortOrder.isEmpty() && sortedWritingEnabled) {
+                Path stagingDirectory = new Path(tempDirectory, "sorting_file_writer_%s_%s".formatted(session.getQueryId(), UUID.randomUUID()));
+                WriteContext writerContext = createWriter(outputPath, partitionData);
+                IcebergFileWriter sortedFileWriter = new IcebergSortingFileWriter(
+                        fileSystem,
+                        stagingDirectory,
+                        writerContext.getWriter(),
+                        sortingFileWriterBufferSize,
+                        sortingFileWriterMaxOpenFiles,
+                        columnTypes,
+                        sortColumnIndexes.build(),
+                        sortOrders.build(),
+                        pageSorter,
```

Comment:
Trino sorting may or may not be same as Iceberg sorting, at least it's not per Iceberg spec.
Actually. I don't think there is any problem with _the code_ here, but I wish Iceberg spec got this fixed: https://github.com/apache/iceberg/pull/2891
cc @RussellSpitzer @yyanyy